### PR TITLE
Clarify docs on implementing Into.

### PR DESCRIPTION
**Pull request by clarfonthey**
This was suggested by @dtolnay in #40380.

This explicitly clarifies in what circumstances you should implement `Into` instead of `From`.

### CODE REVIEW

**Comment by dtolnay** [comment]
This works when I tried it. https://play.rust-lang.org/?gist=29711851d1acb274cd4ba9fd3e05d781

```rust
struct Wrapper(String);

impl From<Wrapper> for String {
    fn from(w: Wrapper) -> String {
        w.0
    }
}

fn main() {}
```

**Comment by Mark-Simulacrum** [comment]
Travis failed:

```
[01:13:49] failures:
[01:13:49]
[01:13:49] ---- convert.rs - convert::Into (line 178) stdout ----
[01:13:49]  thread 'rustc' panicked at 'test compiled while it wasn't supposed to', /checkout/src/librustdoc/test.rs:263
[01:13:49] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:13:49]
[01:13:49] ---- convert.rs - convert::Into (line 194) stdout ----
[01:13:49]  error[E0425]: cannot find value `w` in this scope
[01:13:49]  --> <anon>:7:9
[01:13:49]   |
[01:13:49] 7 |         w.0
[01:13:49]   |         ^ not found in this scope
[01:13:49]
[01:13:49] error[E0053]: method `into` has an incompatible type for trait
[01:13:49]  --> <anon>:6:5
[01:13:49]   |
[01:13:49] 6 | /     fn into(self) -> Wrapper {
[01:13:49] 7 | |         w.0
[01:13:49] 8 | |     }
[01:13:49]   | |_____^ expected struct `std::string::String`, found struct `main::Wrapper`
[01:13:49]   |
[01:13:49]   = note: expected type `fn(main::Wrapper) -> std::string::String`
[01:13:49]              found type `fn(main::Wrapper) -> main::Wrapper`
[01:13:49]
[01:13:49] error: aborting due to previous error(s)
[01:13:49]
[01:13:49] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc/session/mod.rs:214
[01:13:49] failures:
[01:13:49]     convert.rs - convert::Into (line 178)
[01:13:49]     convert.rs - convert::Into (line 194)
[01:13:49]
[01:13:49] test result: FAILED. 1200 passed; 2 failed; 12 ignored; 0 measured; 0 filtered out
```

File: src/libcore/convert.rs [hunk 1]
```diff
@@ -169,6 +169,40 @@ pub trait AsMut<T: ?Sized> {
 /// - [`From<T>`][From]` for U` implies `Into<U> for T`
 /// - [`into`] is reflexive, which means that `Into<T> for T` is implemented
 ///
+/// # Implementing `Into`
+///
+/// There is one exception to implementing `Into`: if the destination type is
+/// not part of the current crate, then you can't implement `From` directly.
+/// For example, take this crate:
+///
+/// ```{.compile_fail}
```
Comment (steveklabnik) [inline src/libcore/convert.rs]: no need for the `{}`s

**Review by steveklabnik — CHANGES_REQUESTED** [review]

**Review by dtolnay — CHANGES_REQUESTED** [review]
The compile_fail code block actually compiles fine.

**Comment by shepmaster** [comment]
@clarcharr a gentle reminder that this PR is awaiting your deft hand to fix the failures and respond to review feedback.

**Comment by clarcharr** [comment]
@shepmaster my deft hand has responded.

**Comment by Mark-Simulacrum** [comment]
Looks like this still needs a few more updates to get Travis passing.

```
[01:04:37] failures:
[01:04:37] 
[01:04:37] ---- convert.rs - convert::Into (line 191) stdout ----
[01:04:37] 	error[E0053]: method `into` has an incompatible type for trait
[01:04:37]  --> <anon>:6:5
[01:04:37]   |
[01:04:37] 6 | /     fn into(self) -> Wrapper<T> {
[01:04:37] 7 | |         self.0
[01:04:37] 8 | |     }
[01:04:37]   | |_____^ expected struct `std::vec::Vec`, found struct `main::Wrapper`
[01:04:37]   |
[01:04:37]   = note: expected type `fn(main::Wrapper<T>) -> std::vec::Vec<T>`
[01:04:37]              found type `fn(main::Wrapper<T>) -> main::Wrapper<T>`
[01:04:37] 
[01:04:37] error: aborting due to previous error(s)
[01:04:37] 
[01:04:37] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc/session/mod.rs:214
[01:04:37] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:04:37] 
[01:04:37] 
[01:04:37] failures:
[01:04:37]     convert.rs - convert::Into (line 191)
[01:04:37] 
[01:04:37] test result: FAILED. 1203 passed; 1 failed; 13 ignored; 0 measured; 0 filtered out
```

**Comment by clarcharr** [comment]
My deft hand failed me once again :(

(fixed the error)

File: src/libcore/convert.rs [hunk 2]
```diff
@@ -169,6 +169,40 @@ pub trait AsMut<T: ?Sized> {
...
 /// - [`into`] is reflexive, which means that `Into<T> for T` is implemented
 ///
+/// # Implementing `Into`
+///
+/// There is one exception to implementing `Into`, and it's kind of esoteric.
+/// If the destination type is not part of the current crate, and it uses a
+/// generic variable, then you can't implement `From` directly.  For example,
+/// take this crate:
+///
+/// ```.compile_fail
```
Comment (Mark-Simulacrum) [inline src/libcore/convert.rs]: No need for the `.` here.

Comment (clarcharr) [inline src/libcore/convert.rs]: Fixed

File: src/libcore/convert.rs (hunk 1 above)
Comment (clarcharr) [inline src/libcore/convert.rs]: I fixed this

**Comment by clarcharr** [comment]
Rebased and made the changes.

**Review by Mark-Simulacrum — APPROVED** [review]

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
@bors: r+ rollup

thanks!

**Pull request merged**